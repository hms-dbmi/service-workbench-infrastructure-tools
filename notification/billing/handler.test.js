const AWS = require('aws-sdk');
const { billingNotification } = require('./handler');
const { byUser, byEnv, none } = require('./handler.test.data');
jest.mock("aws-sdk");

const logger = {};

process.env.PROJECT = 'dev';
process.env.TOPIC_ARN = 'arn:aws:sns:us-east-1:000000000:swb-tools-billing-notification-dev-topic';
process.env.THRESHOLD = 1;
process.env.ENVIRONMENT = 'Dev';

const publishMock = jest.fn();
AWS.SNS.prototype.publish = jest.fn().mockImplementation(params => ({
  promise: () => publishMock(params)
}));

const getCostAndUsageMock = jest.fn();
AWS.CostExplorer.prototype.getCostAndUsage = jest.fn().mockImplementation(params => ({
  promise: () => getCostAndUsageMock(params)
}));

describe('billingNotification', () => {
  beforeEach(() => {
    // capture and print nothing
    logger.info = jest.spyOn(console, "log").mockImplementation(() => { });
    logger.error = jest.spyOn(console, "error").mockImplementation(() => { });
  });
  it('should return status 200 on success', async () => {
    getCostAndUsageMock.mockResolvedValueOnce(none);
    const response = await billingNotification();
    expect(response.statusCode).toEqual(200);
  });
  it('should return status 500 on cost explorer error', async () => {
    getCostAndUsageMock.mockRejectedValueOnce(new Error('Some SNS topic publishing error'));
    const response = await billingNotification();
    expect(response.statusCode).toEqual(500);
  });
  it('should return status 500 on sns topic publish error', async () => {
    publishMock.mockRejectedValueOnce(new Error('Some SNS topic publishing error'));
    const response = await billingNotification();
    expect(response.statusCode).toEqual(500);
  });
  it('should attempt to publish to topic with correct parameters', async () => {
    getCostAndUsageMock.mockResolvedValueOnce(byUser);
    getCostAndUsageMock.mockResolvedValueOnce(byEnv['user-1']);
    getCostAndUsageMock.mockResolvedValueOnce(byEnv['user-2']);
    getCostAndUsageMock.mockResolvedValueOnce(byEnv['user-6']);
    await billingNotification();
    expect(publishMock.mock.calls).toMatchSnapshot();
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
});