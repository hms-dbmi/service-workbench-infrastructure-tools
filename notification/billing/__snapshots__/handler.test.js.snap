// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`billingNotification should attempt to publish to topic with correct parameters 1`] = `
Array [
  Array [
    Object {
      "Message": "Below users are over the $1.00 threshold for 'dev' project:

user-1 - $2.04
  - d7d0a71c-41b4-4e13-89e2-cb05d2fc2559 - $1.98
  - 2133912f-f276-4f6a-a1d3-90d41f3387f6 - $0.06

user-2 - $1.64
  - 11e7843d-d7ea-48ea-8094-639d033352cc - $0.80
  - 92c09776-cca9-4bef-94fe-e5a994f78656 - $0.55
  - 0c990b57-c94d-40a2-8c8d-7d4d67a0a0ed - $0.19
  - 8f89e668-96dd-4e43-949f-1f30db21c9dd - $0.10

user-6 - $1.26
  - e35741d3-eeb0-42f8-bf69-3f2534a1208e - $1.05
  - 36389203-9365-41cd-aa90-8c0cb7e85922 - $0.20",
      "Subject": "Dev SWB - Billing Notice",
      "TopicArn": "arn:aws:sns:us-east-1:000000000:swb-tools-billing-notification-dev-topic",
    },
  ],
]
`;
